src/classes/Collision/CapsuleConvexHull.js

```diff
@@ -0,0 +1,240 @@
+        return closestPointOnHull !== null ?
+            Goblin.Collision.capsuleConvexHull._shallowCapsuleConvexHull( capsule, convexHull, closestPointOnHull, innerSegmentStart, innerSegmentEnd, doLightweightCollision ) :
+            Goblin.Collision.capsuleConvexHull._deepCapsuleConvexHull( capsule, convexHull, doLightweightCollision );
+    };
+} )();
+
+Goblin.Collision.capsuleConvexHull._shallowCapsuleConvexHull = ( function() {
+    var closestPointOnInnerSegment = new Goblin.Vector3();
+    var contactNormal = new Goblin.Vector3();
+    var invertedRotationQuaternion = new Goblin.Quaternion();
+    var referencePlaneNormal = new Goblin.Vector3();
+    var clippedInnerSegmentStart = new Goblin.Vector3();
+    var clippedInnerSegmentEnd = new Goblin.Vector3();
+    var innerSegmentDirection = new Goblin.Vector3();
+    var epsilon = 0.01;
+
+    return function( capsule, convexHull, closestPointOnHull, innerSegmentStart, innerSegmentEnd, doLightweightCollision ) {
+        Goblin.GeometryMethods.findClosestPointOnASegment( innerSegmentStart, innerSegmentEnd, closestPointOnHull, closestPointOnInnerSegment );
+        contactNormal.subtractVectors( closestPointOnHull, closestPointOnInnerSegment );
+        var distanceToClosestPoint = contactNormal.normalize();
+
+        // The capsule overlaps the hull if the distance of the inner capsule segment to the hull
+        // is less or equal than the capsule radius.
+        var penetrationDepth = capsule.shape.radius - distanceToClosestPoint;
+        if ( penetrationDepth < 0 ) {
+            return null;
+        }
+
+        /**
+         * @type {Goblin.ContactDetails}
+         */
+        var contact = Goblin.ObjectPool.getObject( 'ContactDetails' );
+        contact.object_a = capsule;
+        contact.object_b = convexHull;
+        if ( doLightweightCollision ) {
+            contact.is_lightweight = true;
+            return [ contact ];
+        }
+
+        contact.penetration_depth = penetrationDepth;
+        contact.contact_normal.copy( contactNormal );
+
+        innerSegmentDirection.subtractVectors( innerSegmentEnd, innerSegmentStart );
+        innerSegmentDirection.normalize();
+        var dotBetweenInnerSegmentDirectionAndContactNormal = innerSegmentDirection.dot( contactNormal );
+        if ( Math.abs( dotBetweenInnerSegmentDirectionAndContactNormal ) > epsilon ) {
+            Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contact, closestPointOnInnerSegment, closestPointOnHull );
+            return [ contact ];
+        }
+
+        invertedRotationQuaternion.invertQuaternion( convexHull.rotation );
+        invertedRotationQuaternion.transformVector3Into( contactNormal, referencePlaneNormal );
+        referencePlaneNormal.scale( -1 );
+        var referenceFace = null;
+        for ( var i = 0; i < convexHull.shape.faces.length; i++ ) {
```

Comment:
I bet this part will repeat over and over in many places - shouldn't we extact it to a convex hull method?

Reply:
Not sure what part are you referring to, but all the code in ±5 lines from there is completely unique